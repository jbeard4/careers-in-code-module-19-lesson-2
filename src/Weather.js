import React from 'react';
import { fetchWeatherForCurrentLocation } from './fetch-weather';

const weatherRowComponents = [
  ({name}) => <td>{name}</td>,
  ({isDaytime}) => <td>{isDaytime}</td>,
  ({temperature, temperatureUnit}) => <td>{temperature}&deg;{temperatureUnit}</td>,
  ({temperatureTrend}) => <td>{temperatureTrend}</td>,
  ({windSpeed}) => <td>{windSpeed}</td>,
  ({windDirection}) => <td>{windDirection}</td>,
  ({shortForecast, icon}) => <td><img alt={shortForecast} src={icon}/></td>,
  ({shortForecast}) => <td>{shortForecast}</td>,
  ({detailedForecast}) => <td>{detailedForecast}</td>
];

export function WeatherComponent({forecast}){
  return forecast ? 
    <table>
      <tbody>
        {
          weatherRowComponents.map( (C, i) => <tr key={i}>
              {
                forecast.map( period => <C key={period.name} {...period}/> )
              }
            </tr>
          )
        }
      </tbody>
    </table>:
    <p>Loading forecast</p>;
}


export class WeatherContainer extends React.Component {

  constructor(props){
    super(props);
    this.state = {forecast : null};
    fetchWeatherForCurrentLocation().then(forecast => {
      this.setState({ forecast :  forecast.properties.periods })
    }).catch(e => {
      this.setState({ error : e })
    });
  }

  render(){
    return this.state.error ? 
      <div>Error loading weather: {this.state.error.message}</div> :
      <WeatherComponent forecast={this.state.forecast}/>;
  } 
}
