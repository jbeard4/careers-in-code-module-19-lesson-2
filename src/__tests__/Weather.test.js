// Link.react.test.js
import React from 'react';
import { WeatherComponent, WeatherContainer } from '../Weather';
import renderer from 'react-test-renderer';
import forecastMock from './forecast.mock.json';
jest.mock('../fetch-weather');


test('Test WeatherComponent render', () => {
  const component = renderer.create(
    <WeatherComponent forecast={forecastMock.properties.periods}/>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});


test('Test WeatherContainer successful render', () => {
  const component = renderer.create(
    <WeatherContainer />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  return new Promise((resolve, reject) => {
    setTimeout(() => {
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
      resolve()
    }, 10);
  });
});
