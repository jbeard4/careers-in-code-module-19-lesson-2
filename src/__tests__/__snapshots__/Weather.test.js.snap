// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test WeatherComponent render 1`] = `
<table>
  <tbody>
    <tr>
      <td>
        Today
      </td>
      <td>
        Tonight
      </td>
      <td>
        Monday
      </td>
      <td>
        Monday Night
      </td>
      <td>
        Tuesday
      </td>
      <td>
        Tuesday Night
      </td>
      <td>
        Wednesday
      </td>
      <td>
        Wednesday Night
      </td>
      <td>
        Thursday
      </td>
      <td>
        Thursday Night
      </td>
      <td>
        Friday
      </td>
      <td>
        Friday Night
      </td>
      <td>
        Saturday
      </td>
      <td>
        Saturday Night
      </td>
    </tr>
    <tr>
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
    </tr>
    <tr>
      <td>
        82
        °
        F
      </td>
      <td>
        67
        °
        F
      </td>
      <td>
        74
        °
        F
      </td>
      <td>
        62
        °
        F
      </td>
      <td>
        75
        °
        F
      </td>
      <td>
        60
        °
        F
      </td>
      <td>
        74
        °
        F
      </td>
      <td>
        60
        °
        F
      </td>
      <td>
        79
        °
        F
      </td>
      <td>
        63
        °
        F
      </td>
      <td>
        82
        °
        F
      </td>
      <td>
        66
        °
        F
      </td>
      <td>
        84
        °
        F
      </td>
      <td>
        68
        °
        F
      </td>
    </tr>
    <tr>
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
    </tr>
    <tr>
      <td>
        13 to 17 mph
      </td>
      <td>
        3 to 15 mph
      </td>
      <td>
        3 mph
      </td>
      <td>
        2 to 6 mph
      </td>
      <td>
        8 mph
      </td>
      <td>
        2 to 8 mph
      </td>
      <td>
        7 mph
      </td>
      <td>
        3 to 7 mph
      </td>
      <td>
        7 mph
      </td>
      <td>
        6 mph
      </td>
      <td>
        2 to 6 mph
      </td>
      <td>
        1 to 7 mph
      </td>
      <td>
        8 mph
      </td>
      <td>
        3 to 8 mph
      </td>
    </tr>
    <tr>
      <td>
        W
      </td>
      <td>
        W
      </td>
      <td>
        NW
      </td>
      <td>
        NW
      </td>
      <td>
        NW
      </td>
      <td>
        W
      </td>
      <td>
        W
      </td>
      <td>
        SW
      </td>
      <td>
        W
      </td>
      <td>
        SW
      </td>
      <td>
        SW
      </td>
      <td>
        S
      </td>
      <td>
        SW
      </td>
      <td>
        S
      </td>
    </tr>
    <tr>
      <td>
        <img
          alt="Mostly Sunny"
          src="https://api.weather.gov/icons/land/day/sct?size=medium"
        />
      </td>
      <td>
        <img
          alt="Partly Cloudy"
          src="https://api.weather.gov/icons/land/night/sct?size=medium"
        />
      </td>
      <td>
        <img
          alt="Chance Rain Showers"
          src="https://api.weather.gov/icons/land/day/rain_showers,30?size=medium"
        />
      </td>
      <td>
        <img
          alt="Chance Rain Showers"
          src="https://api.weather.gov/icons/land/night/rain_showers,30?size=medium"
        />
      </td>
      <td>
        <img
          alt="Mostly Sunny"
          src="https://api.weather.gov/icons/land/day/sct?size=medium"
        />
      </td>
      <td>
        <img
          alt="Mostly Clear"
          src="https://api.weather.gov/icons/land/night/few?size=medium"
        />
      </td>
      <td>
        <img
          alt="Slight Chance Rain Showers"
          src="https://api.weather.gov/icons/land/day/rain_showers?size=medium"
        />
      </td>
      <td>
        <img
          alt="Slight Chance Rain Showers then Partly Cloudy"
          src="https://api.weather.gov/icons/land/night/rain_showers/sct?size=medium"
        />
      </td>
      <td>
        <img
          alt="Chance Showers And Thunderstorms"
          src="https://api.weather.gov/icons/land/day/tsra_hi,30?size=medium"
        />
      </td>
      <td>
        <img
          alt="Mostly Clear"
          src="https://api.weather.gov/icons/land/night/few?size=medium"
        />
      </td>
      <td>
        <img
          alt="Sunny"
          src="https://api.weather.gov/icons/land/day/few?size=medium"
        />
      </td>
      <td>
        <img
          alt="Partly Cloudy"
          src="https://api.weather.gov/icons/land/night/sct?size=medium"
        />
      </td>
      <td>
        <img
          alt="Mostly Sunny"
          src="https://api.weather.gov/icons/land/day/sct?size=medium"
        />
      </td>
      <td>
        <img
          alt="Partly Cloudy"
          src="https://api.weather.gov/icons/land/night/sct?size=medium"
        />
      </td>
    </tr>
    <tr>
      <td>
        Mostly Sunny
      </td>
      <td>
        Partly Cloudy
      </td>
      <td>
        Chance Rain Showers
      </td>
      <td>
        Chance Rain Showers
      </td>
      <td>
        Mostly Sunny
      </td>
      <td>
        Mostly Clear
      </td>
      <td>
        Slight Chance Rain Showers
      </td>
      <td>
        Slight Chance Rain Showers then Partly Cloudy
      </td>
      <td>
        Chance Showers And Thunderstorms
      </td>
      <td>
        Mostly Clear
      </td>
      <td>
        Sunny
      </td>
      <td>
        Partly Cloudy
      </td>
      <td>
        Mostly Sunny
      </td>
      <td>
        Partly Cloudy
      </td>
    </tr>
    <tr>
      <td>
        Mostly sunny, with a high near 82. West wind 13 to 17 mph.
      </td>
      <td>
        Partly cloudy, with a low around 67. West wind 3 to 15 mph.
      </td>
      <td>
        A chance of rain showers after 10am. Mostly cloudy, with a high near 74. Northwest wind around 3 mph. Chance of precipitation is 30%. New rainfall amounts less than a tenth of an inch possible.
      </td>
      <td>
        A chance of rain showers. Mostly cloudy, with a low around 62. Northwest wind 2 to 6 mph. Chance of precipitation is 30%. New rainfall amounts less than a tenth of an inch possible.
      </td>
      <td>
        Mostly sunny, with a high near 75. Northwest wind around 8 mph.
      </td>
      <td>
        Mostly clear, with a low around 60.
      </td>
      <td>
        A slight chance of rain showers. Mostly sunny, with a high near 74.
      </td>
      <td>
        A slight chance of rain showers before 8pm. Partly cloudy, with a low around 60.
      </td>
      <td>
        A chance of showers and thunderstorms. Mostly sunny, with a high near 79. Chance of precipitation is 30%.
      </td>
      <td>
        Mostly clear, with a low around 63.
      </td>
      <td>
        Sunny, with a high near 82.
      </td>
      <td>
        Partly cloudy, with a low around 66.
      </td>
      <td>
        Mostly sunny, with a high near 84.
      </td>
      <td>
        Partly cloudy, with a low around 68.
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Test WeatherContainer successful render 1`] = `
<p>
  Loading forecast
</p>
`;

exports[`Test WeatherContainer successful render 2`] = `
<table>
  <tbody>
    <tr>
      <td>
        Today
      </td>
      <td>
        Tonight
      </td>
      <td>
        Monday
      </td>
      <td>
        Monday Night
      </td>
      <td>
        Tuesday
      </td>
      <td>
        Tuesday Night
      </td>
      <td>
        Wednesday
      </td>
      <td>
        Wednesday Night
      </td>
      <td>
        Thursday
      </td>
      <td>
        Thursday Night
      </td>
      <td>
        Friday
      </td>
      <td>
        Friday Night
      </td>
      <td>
        Saturday
      </td>
      <td>
        Saturday Night
      </td>
    </tr>
    <tr>
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
    </tr>
    <tr>
      <td>
        82
        °
        F
      </td>
      <td>
        67
        °
        F
      </td>
      <td>
        74
        °
        F
      </td>
      <td>
        62
        °
        F
      </td>
      <td>
        75
        °
        F
      </td>
      <td>
        60
        °
        F
      </td>
      <td>
        74
        °
        F
      </td>
      <td>
        60
        °
        F
      </td>
      <td>
        79
        °
        F
      </td>
      <td>
        63
        °
        F
      </td>
      <td>
        82
        °
        F
      </td>
      <td>
        66
        °
        F
      </td>
      <td>
        84
        °
        F
      </td>
      <td>
        68
        °
        F
      </td>
    </tr>
    <tr>
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
      <td />
    </tr>
    <tr>
      <td>
        13 to 17 mph
      </td>
      <td>
        3 to 15 mph
      </td>
      <td>
        3 mph
      </td>
      <td>
        2 to 6 mph
      </td>
      <td>
        8 mph
      </td>
      <td>
        2 to 8 mph
      </td>
      <td>
        7 mph
      </td>
      <td>
        3 to 7 mph
      </td>
      <td>
        7 mph
      </td>
      <td>
        6 mph
      </td>
      <td>
        2 to 6 mph
      </td>
      <td>
        1 to 7 mph
      </td>
      <td>
        8 mph
      </td>
      <td>
        3 to 8 mph
      </td>
    </tr>
    <tr>
      <td>
        W
      </td>
      <td>
        W
      </td>
      <td>
        NW
      </td>
      <td>
        NW
      </td>
      <td>
        NW
      </td>
      <td>
        W
      </td>
      <td>
        W
      </td>
      <td>
        SW
      </td>
      <td>
        W
      </td>
      <td>
        SW
      </td>
      <td>
        SW
      </td>
      <td>
        S
      </td>
      <td>
        SW
      </td>
      <td>
        S
      </td>
    </tr>
    <tr>
      <td>
        <img
          alt="Mostly Sunny"
          src="https://api.weather.gov/icons/land/day/sct?size=medium"
        />
      </td>
      <td>
        <img
          alt="Partly Cloudy"
          src="https://api.weather.gov/icons/land/night/sct?size=medium"
        />
      </td>
      <td>
        <img
          alt="Chance Rain Showers"
          src="https://api.weather.gov/icons/land/day/rain_showers,30?size=medium"
        />
      </td>
      <td>
        <img
          alt="Chance Rain Showers"
          src="https://api.weather.gov/icons/land/night/rain_showers,30?size=medium"
        />
      </td>
      <td>
        <img
          alt="Mostly Sunny"
          src="https://api.weather.gov/icons/land/day/sct?size=medium"
        />
      </td>
      <td>
        <img
          alt="Mostly Clear"
          src="https://api.weather.gov/icons/land/night/few?size=medium"
        />
      </td>
      <td>
        <img
          alt="Slight Chance Rain Showers"
          src="https://api.weather.gov/icons/land/day/rain_showers?size=medium"
        />
      </td>
      <td>
        <img
          alt="Slight Chance Rain Showers then Partly Cloudy"
          src="https://api.weather.gov/icons/land/night/rain_showers/sct?size=medium"
        />
      </td>
      <td>
        <img
          alt="Chance Showers And Thunderstorms"
          src="https://api.weather.gov/icons/land/day/tsra_hi,30?size=medium"
        />
      </td>
      <td>
        <img
          alt="Mostly Clear"
          src="https://api.weather.gov/icons/land/night/few?size=medium"
        />
      </td>
      <td>
        <img
          alt="Sunny"
          src="https://api.weather.gov/icons/land/day/few?size=medium"
        />
      </td>
      <td>
        <img
          alt="Partly Cloudy"
          src="https://api.weather.gov/icons/land/night/sct?size=medium"
        />
      </td>
      <td>
        <img
          alt="Mostly Sunny"
          src="https://api.weather.gov/icons/land/day/sct?size=medium"
        />
      </td>
      <td>
        <img
          alt="Partly Cloudy"
          src="https://api.weather.gov/icons/land/night/sct?size=medium"
        />
      </td>
    </tr>
    <tr>
      <td>
        Mostly Sunny
      </td>
      <td>
        Partly Cloudy
      </td>
      <td>
        Chance Rain Showers
      </td>
      <td>
        Chance Rain Showers
      </td>
      <td>
        Mostly Sunny
      </td>
      <td>
        Mostly Clear
      </td>
      <td>
        Slight Chance Rain Showers
      </td>
      <td>
        Slight Chance Rain Showers then Partly Cloudy
      </td>
      <td>
        Chance Showers And Thunderstorms
      </td>
      <td>
        Mostly Clear
      </td>
      <td>
        Sunny
      </td>
      <td>
        Partly Cloudy
      </td>
      <td>
        Mostly Sunny
      </td>
      <td>
        Partly Cloudy
      </td>
    </tr>
    <tr>
      <td>
        Mostly sunny, with a high near 82. West wind 13 to 17 mph.
      </td>
      <td>
        Partly cloudy, with a low around 67. West wind 3 to 15 mph.
      </td>
      <td>
        A chance of rain showers after 10am. Mostly cloudy, with a high near 74. Northwest wind around 3 mph. Chance of precipitation is 30%. New rainfall amounts less than a tenth of an inch possible.
      </td>
      <td>
        A chance of rain showers. Mostly cloudy, with a low around 62. Northwest wind 2 to 6 mph. Chance of precipitation is 30%. New rainfall amounts less than a tenth of an inch possible.
      </td>
      <td>
        Mostly sunny, with a high near 75. Northwest wind around 8 mph.
      </td>
      <td>
        Mostly clear, with a low around 60.
      </td>
      <td>
        A slight chance of rain showers. Mostly sunny, with a high near 74.
      </td>
      <td>
        A slight chance of rain showers before 8pm. Partly cloudy, with a low around 60.
      </td>
      <td>
        A chance of showers and thunderstorms. Mostly sunny, with a high near 79. Chance of precipitation is 30%.
      </td>
      <td>
        Mostly clear, with a low around 63.
      </td>
      <td>
        Sunny, with a high near 82.
      </td>
      <td>
        Partly cloudy, with a low around 66.
      </td>
      <td>
        Mostly sunny, with a high near 84.
      </td>
      <td>
        Partly cloudy, with a low around 68.
      </td>
    </tr>
  </tbody>
</table>
`;
